docs/coding-guidelines/breaking-change-rules.md

```diff
@@ -169,6 +169,8 @@ Breaking Change Rules
 
 * Changing a `struct` type to a `ref struct` type and vice versa
 
+* Adding a new `ref` field to a type that didn't have any `ref` field before (recursively)
```

Comment:
I am not following what's the rationale for this. How is this different from adding/removing any fields? We have no guarantees for internal type layout compatibility, unless the guarantee is explicitly documented. 

"Changing a `struct` type to a `ref struct` type and vice versa" is very different - it is very obvious compile time breaking change.

Reply:
I think the point of it is that now add a `ref` field could break someone if they were using the type in an `Explicit` offset scenario. For example, the BCL has some `ref struct` that just has an `int `field. A user consumes that type in a user defined type as an `Explicit` field and the containing library is published to NuGet. In the next major version a `ref int` field is added. When that NuGet package is consumed on a new BCL, the library fails to load because the `ref int` field forces a new alignment restriction.

By definition this would now break. We can argue if it matters, but it would cause a failure at type load time.

Reply:
> I think the point of it is that now adding a ref field could break someone if they were using the type in an Explicit offset scenario. 

It is the case for any field changes in public BCL value types. You can get the same break by adding/removing `int` or `object` fields too. There is nothing special about ref fields here.

Field layout of BCL value types is not a public contract that people can depend on, unless it is explicitly documented for specific types.

Reply:
> "Changing a struct type to a ref struct type and vice versa" is very different - it is very obvious compile time breaking change.

> It is the case for any field changes in public BCL value types. You can get the same break by adding/removing int or object fields too. There is nothing special about ref fields here.

Between the two statements above, it looks like the distinction in this case is compile time vs run time failure. Is that the conversation we are having? In the sense that compile time failure is not acceptable, but run time failure is?

Reply:
There happens to be a compile time vs. runtime time difference between these two case, but that's just accidental. Runtime breaking changes are not acceptable either in general.

I think that the discussion that we are having here is about distinction between what we define as a breaking change vs. change that can break apps that depend on undocumented behaviors.

Reply:
> I think that the discussion that we are having here is about distinction between what we define as a breaking change vs. change that can break apps that depend on undocumented behaviors.

Okay, Thanks.

> Field layout of BCL value types is not a public contract that people can depend on, unless it is explicitly documented for specific types.

Is this explicitly documented somewhere?

Reply:
> Is this explicitly documented somewhere?

I am not sure whether it is mentioned in the official docs. Should we mention it in https://learn.microsoft.com/en-us/dotnet/standard/native-interop/best-practices ?

This falls into the same category as dangers of private reflection that are not particularly well documented either.

Reply:
> Should we mention it in https://learn.microsoft.com/en-us/dotnet/standard/native-interop/best-practices ?

Seems like a reasonable place to mention it. As you mentioned this is also a refelction story. Is there a general place for reflection or type system details that you'd recommend?

Reply:
> It is the case for any field changes in public BCL value types. You can get the same break by adding/removing `int` or `object` fields too. There is nothing special about ref fields here.

I thought we already consider such things breaking as well - for example adding an object typed field to a struct makes the struct no longer meet `unmanaged` constraint. I recall Roslyn team even did work to put a dummy object filed into such struct for `ref` assemblies because these non-contractual private implementation details actually matter.

Reply:
If the libraries folks agree, I think it would be fine to explicitly match the current Roslyn behavior and say that adding first field of a new kind to an existing value type is disallowed change. Kind can be any of objref, byref or generic type argument, and the rule applies recursively. 

Introducing new significant state in existing value types is close to impossible to do in compatible way for other reasons (e.g. it is why we have both DateTime and DateTimeOffset), so these situations do not come up in practice. I guess it is the reason why nobody bothered to mention it in the breaking changes doc up until now.

Reply:
I will create a discussion issue for this topic and loop in libraries folks. Updating the markdown is a cheaper PR than continually running the build/test legs.

Reply:
> This falls into the same category as dangers of private reflection that are not particularly well documented either.

I have created https://github.com/dotnet/runtime/issues/112114

Reply:
> Should we mention it in https://learn.microsoft.com/en-us/dotnet/standard/native-interop/best-practices

Opened https://github.com/dotnet/docs/pull/44672